src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java

```diff
@@ -119,10 +133,76 @@ public TypeName getTypeNameForElementType(TypeRef type) {
 
   @Override
   public TypeName getTypeName(ProtoElement elem) {
-    if (elem.getFullName().equals("google.protobuf.Empty")) {
-      return typeNameConverter.getTypeName("\\google\\protobuf\\EmptyC");
+    try {
+      return getTypeName(elem, MAX_NESTED_DEPTH);
+    } catch (IllegalStateException e) {
+      throw new IllegalStateException("Could not determine type name for elem: " + elem, e);
+    }
+  }
+
+  /**
+   * This function recursively determines the PHP type name for a proto message. Recursion is
+   * required because in PHP nested messages are handled differently from non-nested messages. For
+   * example, the following proto definition: <code>
+   * package example;
+   * message Top {
+   *  message Nested {
+   *    message DeepNested {}
+   *  }
+   * }
+   * </code> will generate these PHP types:
+   *
+   * <ul>
+   *   <li>\Example\Top
+   *   <li>\Example\Top_Nested
+   *   <li>\Example\Top_Nested_DeepNested
+   * </ul>
+   *
+   * <p>To correctly output these type names, we need to check whether the parent of a proto element
+   * is a message, and if so use '_' as a separator.
+   */
+  private TypeName getTypeName(ProtoElement elem, int maxDepth) {
```

Comment:
You can remove the maxDepth parameter and use MAX_NESTED_DEPTH directly in this method.